### PR TITLE
[pvr][guilib][json][skin.estuary][skin.estouchy] Add new API HasArchive for ListItem

### PR DESCRIPTION
**Pull request by arthur-liberman**
## Description
This PR adds a new API for ListItem. The skin can now query the channel or the program in the EPG whether it can be "Played from EPG tag" or not (introduced in 21900fd). 
I'm not sure about the naming used as "archive", and if you can think of a better naming scheme for this I'm all for it.

## Motivation and Context
In some cases this functionality is not supported on all available channels, and it is very useful to have a visual indication of whether a channel/program supports it or not. 

## Screenshots (if appropriate):
image
image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
Tusks for this nice contribution.

First, as this PR includes  PVR add-on API, JSON API and database version bumps it can't make it into v18, sorry. We're nearing RC phase and incompatible API changes are not possible at this stage. So, this one is for v19.

Code looks pretty much okay, I will do a review later.

I'm also uncertain about the wording - we can figure this out later.

Questions:
- can channels without EPG have "archive" support? 
- which pvr add-on did you extend to test your charges / make the screen shots? Show me the code, please.
- the green dots on the epg grid - this is "ListItem.IsPlayable ", right?

File: xbmc/interfaces/json-rpc/schema/types.json [hunk 1]
```diff
@@ -1016,7 +1016,8 @@
       "icon": { "type": "string" },
       "channelnumber": { "type": "integer" },
       "subchannelnumber": { "type": "integer" },
-      "isrecording": { "type": "boolean" }
+      "isrecording": { "type": "boolean" },
+      "hasarchive": { "type": "boolean" }
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/types.json]: Thanks for updating the API as part of this feature, and for correctly adding the field at the end of the type.

However addition of fields to the JSON schema requires a bump to the _minor_ version number (the middle one) in https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/version.txt as a non-breaking change, communicating to API customers that things have changed. Adding [JSON] to the commit comment would also make it easier to spot.

**Comment by arthur-liberman** [comment]
@ksooo 
Some services refer to this as "catchup", but I think Archive is still a very common term for it.
* Channels without EPG can have archive support. This will depend on the PVR add-on implementation of course, but as long as the channel has support for it, you can timeshift "blind", without knowing exactly on what show you land.
* I extended (and renamed) pvr.iptvsimple, but it includes more functionality than just support for this. It's currently in patch form, and I have a PR open for it here: coreelec/coreelec#138
I did test Kodi with just the changes from this PR (HasArchive) with it, and it worked as expected.
* I tried ListItem.IsPlayable, that did nothing, I don't think there is an implementation for it as I did a quick search for it. The green dots is the new "ListItem.HasArchive", I simply added a new png to the Confluence skin and a new image element, it only shows the green dot on shows that are "IsPlayable" in their GetEPGInfoTag + not recording, same for channels.

File: xbmc/interfaces/json-rpc/schema/types.json (hunk 1 above)
Comment (arthur-liberman) [reply xbmc/interfaces/json-rpc/schema/types.json]: Thanks. I'll do that once I'm home.
Should I make another commit, or amend the previous one and force push? I'm not yet familiar with how PRs to Kodi work.

Comment (MartijnKaijser) [reply xbmc/interfaces/json-rpc/schema/types.json]: You can amend and force push changes. You can just do Fixup commit and in the end squash them into clear separated changes.

**Comment by ksooo** [comment]
> I extended (and renamed) pvr.iptvsimple, but it includes more functionality than just support for this. It's currently in patch form, and I have a PR open for it here: CoreELEC/CoreELEC#138 (oops, I'll need to update that PR with the changes for the HasArchive flag)

I'm STRONGLY against extending random distros with heavily patched versions of official Kodi addons and Kodi itself! If you want to do this kind of stuff you have to rebrand Kodi and the addon.

**Comment by ksooo** [comment]
I just noticed that you renamed the addon. Doesn't it make better, imo.

Why are you doing PRs for CoreELEC instead for xbmc / kodi-pvr ? Don't you want your features it in the upstream products?

**Comment by vpeter4** [comment]
Addon is already using IsEPGTagPlayable.

**Comment by arthur-liberman** [comment]
Ah yes, that works. I meant that it's not implemented for ListItem, for the skin to use as an indication of whether the item is playable or not.

**Comment by ksooo** [comment]
> Ah yes, that works. I meant that it's not implemented for ListItem, for the skin to use as an indication of whether the item is playable or not.

 Ah, right. I thought we have "ListItem.IsPlayable". I was wrong on that.

**Comment by Raybuntu** [comment]
> > I extended (and renamed) pvr.iptvsimple, but it includes more functionality than just support for this. It's currently in patch form, and I have a PR open for it here: [CoreELEC/CoreELEC#138](https://github.com/CoreELEC/CoreELEC/pull/138) (oops, I'll need to update that PR with the changes for the HasArchive flag)
> 
> I'm STRONGLY against extending random distros with heavily patched versions of official Kodi addons and Kodi itself! If you want to do this kind of stuff you have to rebrand Kodi and the addon.

So pvr.iptvsimple is your trademark as well?

**Comment by ksooo** [comment]
> There is currently no way to play a tag from EPG in "Live TV" mode. [...] But really, the user experience with this particular change, where the program simply "timeshifts" back, instead of playing as a "video" is really great.

 Have you tried skip+/skip- while playing Live TV using a PVR addon that properly implements timeshifting? => https://github.com/xbmc/xbmc/pull/14387

**Comment by arthur-liberman** [comment]
Yes, I just found out about it a couple of days ago, actually. It works perfect.

**Comment by ksooo** [comment]
> So pvr.iptvsimple is your trademark as well?

 IANAL, but this is not about trademarks. ID change is about working versioning and updates for the official addon. Name change is about avoiding confusion for users and people trying to give support for the official addon (and the "forked" addon).

**Comment by MartijnKaijser** [comment]
Just a heads up: Any further non constructive comments will be removed without questioning.

**Comment by ksooo** [comment]
> Yes, I just found out about it a couple of days ago, actually. It works perfect.

 So, I guess, for best user experience you should implement proper timeshifting for your addon - no special hacks needed. There are "templates" for this (local ts buffer if no serverside ts) - for example pvr.vuplus has it.

But this is ofc not scope of this PR and should not be discussed further here, it just came to my mind.

**Comment by ksooo** [comment]
> So, I guess, for best user experience you should implement proper timeshifting for your addon - no special hacks needed.

 If I understand the coreelec patch halfway right, then implementing timeshifting support for official pvr.iptvsimple would be the way to go as it will give the users what you want to achieve with pvr.iptvarchive + Kodi patches, but in a clean and acceptable way. What do you think? You could for instance ask @phunkyfish if you need help with that. He just did it for pvr.vuplus. Users will love you if you extend pvr.iptvsimple with proper timeshifting, btw. ;-)

 So, back to this PR, which is generally fine as I said before... I will take a closer look at the code changes the next days.

**Comment by arthur-liberman** [comment]
@ksooo Is there a place where we can discus this further? One obstacle I hit was DVDFactoryInputStream.
m3u8 based playlists are relegated to CDVDInputStreamFFmpeg which doesn't even support the PVR methods. I think with a little bit of cooperative work, on both Kodi and iptvsimple side, we can get this working better than it is working right now with the patches.

**Comment by ksooo** [comment]
I'm not very familiar with videoplayer internals. @FernetMenta what's your take on this?

**Comment by phunkyfish** [comment]
@ksooo @arthur-liberman  Happy to help with timeshifting for IPTV simple.

File: xbmc/interfaces/json-rpc/schema/types.json (hunk 1 above)
Comment (arthur-liberman) [reply xbmc/interfaces/json-rpc/schema/types.json]: Done.

**Comment by arthur-liberman** [comment]
@phunkyfish The Archive feature allows us to have 'pseudo-timeshift'. Basically it works like this:
IPTV providers have an API for accessing the archive. Some build it right into the playlist, using the `catchup-source` #EXTINF tag. There is a separate URL to access the channel at a specific time stamp using a placeholder for the start time of the program in UTC.
There is more information in those tags, such as how many days back you can access on said channel. Others require you to append parameters to the channel URL, such as `?utc={start-time-utc-timestamp}&lutc={current-utc-timestamp}`. This format is supported by many players, such as OTT Player.
Now, to timeshift, all we need is to know the time we want to start watching, and restart the stream with a new timestamp in the URL.
This also allows you to have a timeshift buffer that is as long as the whole archive duration, because nothing is physically stored on your device. With #14387 you can skip whole programs back/forward with ease.

Supporting archive/timeshift in playlists that have the catchup-source url built in is easy, because it's provided to us. The question then becomes what to do with the services that require URL parameters, because these formats can differ from one service to the next. I think that for starters, adding support for `catchup` should be the first step, and we can figure out what to do with the services that require parameters later.

**Comment by phunkyfish** [comment]
Just let me know when/if I can help and happy to do so.

**Comment by vpeter4** [comment]
@phunkyfish: How can we get in touch with you? IRC/....

**Comment by ksooo** [comment]
Can you guys please move the discussion not related to this PR elsewhere. Otherwise I'm loosing track here. Thanks.

**Comment by ksooo** [comment]
I suggest to open a thread in the kodi forum.

**Comment by arthur-liberman** [comment]
Alright, let's get this PR back on topic. 
I pushed changes to iptvsimple and to the confluence skin that will allow everyone to test this PR.
https://github.com/arthur-liberman/pvr.iptvsimple/commit/a7b2ba66ad237720a4f515020f1e34d2a79f4c1c
https://github.com/arthur-liberman/skin.confluence/commit/47e1c96318c6087547ad78be024ba7bdf1042f1d

File: xbmc/pvr/PVRGUIInfo.cpp [hunk 2]
```diff
@@ -1061,6 +1061,18 @@ bool CPVRGUIInfo::GetListItemAndPlayerBool(const CFileItem *item, const CGUIInfo
...
   {
+    case LISTITEM_HASARCHIVE:
+      if (item->IsPVRChannel())
+      {
+        bValue = item->GetPVRChannelInfoTag()->HasArchive();
+        return true;
+      }
+      else if (item->IsEPG())
+      {
+        bValue = item->GetEPGInfoTag()->IsPlayable();
```
Comment (ksooo) [inline xbmc/pvr/PVRGUIInfo.cpp]: As an EPG tag semantically cannot have an archive, I would suggest to introduce a new guiinfo label `ListItem.IsPlayable` and to remove the EPG part from implementation of `Listitem.HasArchive`.

File: xbmc/GUIInfoManager.cpp [hunk 3]
```diff
@@ -3944,6 +3944,7 @@ const infomap listitem_labels[]= {{ "thumb",            LISTITEM_THUMB },
                                   { "hastimerschedule", LISTITEM_HASTIMERSCHEDULE },
                                   { "hasrecording",     LISTITEM_HASRECORDING },
                                   { "isrecording",      LISTITEM_ISRECORDING },
+                                  { "hasarchive",       LISTITEM_HASARCHIVE },
```
Comment (ksooo) [inline xbmc/GUIInfoManager.cpp]: Please add documentation for the new listitem above the `listitem_labels` table.

https://github.com/xbmc/xbmc/pull/14736/files#diff-127e5f2786eb19ac8e0e2f04b3bf8306R3797

File: xbmc/pvr/channels/PVRChannel.h [hunk 4]
```diff
@@ -444,6 +449,7 @@ namespace PVR
     time_t           m_iLastWatched;            /*!< last time channel has been watched */
     bool             m_bChanged;                /*!< true if anything in this entry was changed that needs to be persisted */
     CPVRChannelNumber m_channelNumber;         /*!< the number this channel has in the currently selected channel group */
+    bool             m_bHasArchive;             /*!< true if this channel supports archive */
```
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannel.h]: Could you please elaborate here what an archive is in this context.

**Review by ksooo — COMMENTED** [review]
some minors only.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h [hunk 5]
```diff
@@ -402,6 +402,7 @@ extern "C" {
     unsigned int iEncryptionSystem;                                    /*!< @brief (optional) the encryption ID or CaID of this channel */
     char         strIconPath[PVR_ADDON_URL_STRING_LENGTH];             /*!< @brief (optional) path to the channel icon (if present) */
     bool         bIsHidden;                                            /*!< @brief (optional) true if this channel is marked as hidden */
+    bool         bHasArchive;                                          /*!< @brief (optional) true if this channel supports archive */
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: Could you please elaborate here what an archive is in this context.

Comment (arthur-liberman) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: Ok, will do. I think the way you put it "server-side back buffer" is probably a good description.
Maybe the whole "archive" part should be changed to something else? I couldn't come up with a better name for it than this, but I'm open to suggestions.

File: xbmc/pvr/PVRGUIInfo.cpp (hunk 2 above)
Comment (arthur-liberman) [reply xbmc/pvr/PVRGUIInfo.cpp]: Changed

File: xbmc/GUIInfoManager.cpp (hunk 3 above)
Comment (arthur-liberman) [reply xbmc/GUIInfoManager.cpp]: Done

**Comment by arthur-liberman** [comment]
I have setup branches with updated Confluence skin and pvr.iptvsimple to enable testing this PR.
https://github.com/arthur-liberman/xbmc/commits/add-listitem-hasarchive
https://github.com/arthur-liberman/skin.confluence/commits/archive

File: xbmc/GUIInfoManager.cpp (hunk 3 above)
Comment (ksooo) [reply xbmc/GUIInfoManager.cpp]: Thanks.

File: xbmc/pvr/PVRGUIInfo.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIInfo.cpp]: 👍

**Comment by arthur-liberman** [comment]
It just occurred to me that Estuary is a part of the XBMC repo. I added HasArchive and IsPlayable support to it. Now this PR can be tester more easily.
I have setup a branch with an updated pvr.iptvsimple to enable testing this PR.
https://github.com/arthur-liberman/pvr.iptvsimple/commits/archive

**Comment by arthur-liberman** [comment]
Is there any status updates on this PR? I saw that Kodi has went through some restructuring of the PVR, but after fixing conflicts, this PR still works as expected.

**Comment by ksooo** [comment]
I most probably will cherry pick this PR for the next PVR Addon API change PR I'm currently preparing for v19.

**Comment by arthur-liberman** [comment]
Superseded by #15953

**Pull request closed without merging**